From c8b7e9c2ae4951400c2d8a39a10a84d21c4c5c0b Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megi@xff.cz>
Date: Sun, 26 Jul 2026 00:56:38 +0200
Subject: [PATCH] mfd: core: Don't overwrite sub-device DMA masks with unset
 parent masks

platform_device_alloc() sets up default DMA masks for the sub-device,
but mfd_add_device() then overwrites them with the parent's, which are
NULL/0 for e.g. I2C devices. OF-bound sub-devices (cells with
of_compatible, like rk818-charger) then warn during probe:

  platform rk818-charger.3.auto: DMA mask not set

Inherit the parent's masks only when the parent has them.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/mfd/mfd-core.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/mfd/mfd-core.c b/drivers/mfd/mfd-core.c
index 7aa32b90cf1e..dac085029682 100644
--- a/drivers/mfd/mfd-core.c
+++ b/drivers/mfd/mfd-core.c
@@ -180,9 +180,12 @@ static int mfd_add_device(struct device *parent, int id,
 
 	pdev->dev.parent = parent;
 	pdev->dev.type = &mfd_dev_type;
-	pdev->dev.dma_mask = parent->dma_mask;
-	pdev->dev.dma_parms = parent->dma_parms;
-	pdev->dev.coherent_dma_mask = parent->coherent_dma_mask;
+	if (parent->dma_mask)
+		pdev->dev.dma_mask = parent->dma_mask;
+	if (parent->dma_parms)
+		pdev->dev.dma_parms = parent->dma_parms;
+	if (parent->coherent_dma_mask)
+		pdev->dev.coherent_dma_mask = parent->coherent_dma_mask;
 
 	ret = regulator_bulk_register_supply_alias(
 			&pdev->dev, cell->parent_supplies,
-- 
2.43.0

